In [122]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [123]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
df_lending_data = pd.read_csv("../Credit_Risk/Resources/lending_data.csv")

# Review the DataFrame
df_lending_data

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [124]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending_data["loan_status"]

# Separate the X variable, the features
X = df_lending_data.drop(columns="loan_status")

In [125]:
# Review the y variable Series
y

0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64

In [126]:
# Review the X variable DataFrame
X

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
0                    5                 1       22800  
1                    3                 0       13600  
2                    3                 0       16100  
3                    5                 1       22700  
4                    5                 1       23000  
...                ...               ...         ...  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300  

[77536 rows x 7 columns]

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [127]:
# Check the balance of our target values
y.value_counts()

0    75036
1     2500
Name: loan_status, dtype: int64

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [128]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X,
                                                    y,
                                                    random_state=1,
                                                    stratify=y)
X_train.shape

(58152, 7)

---

In [129]:
# Get count of training set
len(X_train)

58152

In [130]:
# Display training set
X_train.head()

loan_size  interest_rate  borrower_income  debt_to_income  \
44997     8500.0          6.722            43900        0.316629   
13468    10000.0          7.360            49900        0.398798   
52163     8400.0          6.708            43700        0.313501   
5330     11000.0          7.779            53800        0.442379   
9982      8900.0          6.924            45800        0.344978   

       num_of_accounts  derogatory_marks  total_debt  
44997                3                 0       13900  
13468                4                 0       19900  
52163                3                 0       13700  
5330                 5                 1       23800  
9982                 3                 0       15800

In [131]:
# Get count of test set
len(X_test)

19384

In [132]:
# Display test set
X_test.head()

loan_size  interest_rate  borrower_income  debt_to_income  \
36831     9000.0          6.960            46100        0.349241   
75818    16300.0         10.064            75300        0.601594   
36563    12000.0          8.240            58100        0.483649   
13237    10400.0          7.558            51700        0.419729   
43292     7900.0          6.478            41600        0.278846   

       num_of_accounts  derogatory_marks  total_debt  
36831                3                 0       16100  
75818               10                 2       45300  
36563                6                 1       28100  
13237                4                 1       21700  
43292                2                 0       11600

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [133]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
lrm = LogisticRegression(random_state=1)

# Fit the model using training data
lr_model = lrm.fit(X_train, y_train)
lr_model

LogisticRegression(random_state=1)

In [134]:
# Get scores for train and test data
print(lr_model.score(X_train, y_train))
print(lr_model.score(X_test, y_test))

0.9914878250103177
0.9924164259182832


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [135]:
# Make a prediction using the testing data
testing_predictions = lr_model.predict(X_test)
testing_results = pd.DataFrame({"Testing Prediction": testing_predictions, "Actual": y_test})
testing_results

Testing Prediction  Actual
36831                   0       0
75818                   0       1
36563                   0       0
13237                   0       0
43292                   0       0
...                   ...     ...
38069                   0       0
36892                   0       0
5035                    0       0
40821                   0       0
35030                   0       0

[19384 rows x 2 columns]

In [136]:
df_lending_data.shape

(77536, 8)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [137]:
# Print the balanced_accuracy score of the model
from sklearn.metrics import accuracy_score
balanced_accuracy_score(y_test, testing_predictions)

0.9442676901753825

In [138]:
# Generate a confusion matrix for the model
from sklearn.metrics import confusion_matrix
testing_matrix = confusion_matrix(y_test, testing_predictions)
testing_matrix

array([[18679,    80],
       [   67,   558]], dtype=int64)

In [139]:
# Print the classification report for the model
testing_report = classification_report(y_test, testing_predictions)
print(testing_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** `This model resulted in 100% precision for healthy loans (the "0" label) perfectly.  However, for high-risk loans (the "1" label), the precision is 87% indicating that the model is not as good at predicting high-risk loans. Furthermore, there were 67 instances which were FN (False Negatives) indicating that there were 67 loans that were predicted to be high-risk, but were healthy loans.`

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [140]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
oversample = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_train_resampled, y_train_resampled = oversample.fit_resample(X_train, y_train) 

In [141]:
# Count the distinct values of the resampled labels data
y_train_resampled.value_counts()

0    56277
1    56277
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [142]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
resampled_lr_model = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
resampled_lr_model.fit(X_train_resampled, y_train_resampled)

# Make a prediction using the testing data
resampled_predictions = resampled_lr_model.predict(X_test)
resampled_results = pd.DataFrame({"Resampled Testing Prediction:": resampled_predictions, "Actual": y_test})
resampled_results

Resampled Testing Prediction:  Actual
36831                              0       0
75818                              1       1
36563                              0       0
13237                              0       0
43292                              0       0
...                              ...     ...
38069                              0       0
36892                              0       0
5035                               0       0
40821                              0       0
35030                              0       0

[19384 rows x 2 columns]

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [143]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_score(y_test, resampled_predictions)

0.9959744975744975

In [144]:
# Generate a confusion matrix for the model
resampled_matrix = confusion_matrix(y_test, resampled_predictions)
resampled_matrix

array([[18668,    91],
       [    2,   623]], dtype=int64)

In [145]:
# Print the classification report for the model
resampled_report = classification_report(y_test, resampled_predictions)
print(resampled_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      1.00      0.93       625

    accuracy                           1.00     19384
   macro avg       0.94      1.00      0.96     19384
weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** `This model also resulted in 100% precision for healthy loans, but increased in FP (False positives) from 80 to 91.  As it relates to high-risk loans, this model reduced the number of FN (false negatives) from 67 to 2.  Also, the f1-score for high-risk loans went up from .88 to .93 making oversampling an improvement in predicting high-risk loans.  However, with this model, there is a higher chance of wrongly predicting a healthy loan since FP increased from 80 in the original model to 91 in the oversampled model.`